crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -411,42 +476,176 @@ impl Object {
+                    // This is because an object can be extant at some target, but be indexed with
+                    // at a checkpoint_sequence_number <= target.
+                    if let Some(checkpoint_sequence_number) = checkpoint_sequence_number {
+                        // A validation check is not needed as we coalesce both sides to 0 for null
+                        // values, and if left > right, the db will return 0 rows. This also saves
+                        // us a roundtrip.
+                        historical_query = historical_query.filter(
+                            history::checkpoint_sequence_number
+                                .nullable()
+                                .between(left.single_value(), checkpoint_sequence_number),
+                        );
+
+                        snapshot_query = snapshot_query.filter(
+                            snapshot::checkpoint_sequence_number.le(checkpoint_sequence_number),
+                        );
+                    } else {
+                        let right = checkpoints::checkpoints
+                            .select(checkpoints::sequence_number)
+                            .order(checkpoints::sequence_number.desc())
+                            .limit(1);
+
+                        historical_query =
+                            historical_query.filter(history::checkpoint_sequence_number.between(
+                                coalesce(left.single_value(), 0),
+                                coalesce(right.single_value(), 0),
+                            ));
+                    }
+
+                    snapshot_query.union(historical_query)
                 })
                 .optional()
             })
-            .await
-            .map_err(|e| Error::Internal(format!("Failed to fetch object: {e}")))?;
+            .await?;
 
-        stored_obj.map(Self::try_from).transpose()
+        // For the moment, if the object existed at some point, it will have eventually be written to objects_snapshot.
+        // Therefore, if both results are None, the object has never existed.
+        let Some(stored_objs) = results else {
+            return Ok(None);
+        };
+
+        match stored_objs
+            .iter()
+            // First, filter the objects based on the version and checkpoint_sequence_number if provided
+            .filter(|stored_obj| {
+                version.map_or(true, |version| stored_obj.object_version == version)
+                    && checkpoint_sequence_number.map_or(true, |checkpoint| {
+                        stored_obj.checkpoint_sequence_number <= checkpoint
+                    })
+            })
+            // Then, find the object with the largest checkpoint_sequence_number among those
+            // filtered - it should still be within the available range as the db query was bounded
+            .max_by_key(|stored_obj| stored_obj.checkpoint_sequence_number)
```

Comment:
order by object_version instead b/c an object is possible to be updated multiple times in one checkpoint?

Reply:
look it twice, it would make a real diff b/c of order by object version desc and limit 1 above, but ordering by object version desc. is usually more robust.

Reply:
Ah right. This scenario only really happens when we have a valid result from both objects_snapshot and objects_history, most likely to happen when a historical query is made on just object_id and checkpoint_sequence_number. Then yes, we should instead select the object at the larger version, instead of checkpoint_sequence_number.

Reply:
Okay, actually, I suppose there is an edge case here, where objects_snapshot is at checkpoint 0 (hasn't snapshotted anything yet). This is the typical scenario in our e2e tests, since all the test behavior would occur within the first few checkpoints.

Reply:
looks like we also need to coalesce the lhs to default to 0 if objects_snapshot has not taken a snapshot yet